bayesian_changepoint_detection_experiment/01_change_point_detection.py

```diff
@@ -0,0 +1,248 @@
+subsetted_tokens.head()
+
+# ## Calculate percent change for frequency
+
+all_word_pct_change_df = (
+    all_word_freq_df
+    >> ply.query(f"tok in {token_filter_list}")
+    >> ply.group_by("tok")
+    >> ply.arrange("year")
+    >> ply.define(freq_pct_change="frequency.pct_change()")
+    >> ply.ungroup()
+    >> ply.call(".dropna")
+    >> ply.define(year=lambda x: x["year"].apply(lambda y: f"{int(y)-1}-{y}"))
+)
+all_word_pct_change_df.head()
+
+# # Grab semantic change values
+
+distance_files = list(
+    Path("../multi_model_experiment/output/combined_inter_intra_distances").rglob(
+        "saved_*_distance.tsv"
+    )
+)
+print(len(distance_files))
+
+year_distance_map = {
+    re.search(r"\d+", str(year_file)).group(0): (pd.read_csv(str(year_file), sep="\t"))
+    for year_file in tqdm.tqdm(distance_files)
+}
+
+full_token_set_df = pd.concat(
+    [
+        year_distance_map[year] >> ply.query("year_2-year_1 == 1")
+        for year in tqdm.tqdm(year_distance_map)
+    ]
+)
+print(full_token_set_df.shape)
+full_token_set_df.head()
+
+# # Combine both information
+
+# Using an idea similar to SCAF \[[1](https://doi.org/10.1007/s00799-019-00271-6)\], I'm combining the global qst metric with the percent change in frequency.
+# SCAF uses percent change for both metrics; however, the caveat is that their method loses information for the first two timepoints.
+# In my case given that global_distance_qst is a metric that's bound between 0 and 1 frequency percent change can be used directly.
+# By combining these two terms I can use this metric as a means to estimate change and allow for bayesian changepoint detection to calculate the probability of a timepoint change.
+
+merged_frequency_df = (
+    full_token_set_df
+    >> ply_tdy.unite("year", "year_1", "year_2", sep="-")
+    >> ply.inner_join(all_word_pct_change_df, on=["tok", "year"])
+    >> ply.rename(year_pair="year")
+    >> ply.select(
+        "tok",
+        "original_global_distance",
+        "global_distance_qst",
+        "year_pair",
+        "frequency",
+        "freq_pct_change",
+    )
+)
+merged_frequency_df
+
+change_metric_df = (
+    merged_frequency_df
+    >> ply.group_by("tok")
+    >> ply.arrange("year_pair")
+    >> ply.define(change_metric="global_distance_qst + freq_pct_change")
+    >> ply.ungroup()
+)
+change_metric_df
+
+# # Change point detection
+
+# Use Semantic Change Analysis with Frequency (SCAF) to perform bayesian change point detection.
+
+if not Path("output/bayesian_changepoint_data.tsv", sep="\t").exists():
+    change_point_results = []
+    for tok, tok_series_df in tqdm.tqdm(change_metric_df.groupby("tok")):
+        change_metric = np.insert(tok_series_df >> ply.pull("change_metric"), 0, 0)
+        Q, P, Pcp = offcd.offline_changepoint_detection(
+            change_metric,
+            partial(offcd.const_prior, l=(len(change_metric) + 1)),
+            offcd.gaussian_obs_log_likelihood,
```

Comment:
>do you expect the errors to be gaussian? Maybe I'm misinterpreting the model (I'm not very familiar with change-point detection), but is this stating that the distributional assumption of the underlying linear model is Gaussian? I'm just wondering since I think you are working with frequencies (0-1).

Correct. Bayesian detection assumes the likelihood of the data follows a gaussian distribution. I'm actually working with percent change in frequency rather than frequencies on there own. The range here is actually -infty to infty, which means gaussian would be most appropriate for this case.

>Is there any goodness-of-fit metrics provided by this package? plotting them or making a table could be useful.

Sadly this package only provides the estimates. No goodness-of-fit metrics provided.